zio-http/shared/src/main/scala/zio/http/endpoint/Endpoint.scala

```diff
@@ -367,66 +363,69 @@ final case class Endpoint[PathInput, Input, Err, Output, Auth <: AuthType](
+    inCodec(HttpCodec.content[Input2](name))
 
   /**
    * Returns a new endpoint derived from this one, whose request content must
    * satisfy the specified schema and is documented.
    */
-  def in[Input2: HttpContentCodec](name: String, doc: Doc)(implicit
+  def in[Input2: EnvironmentTag: HttpContentCodec](name: String, doc: Doc)(implicit
     combiner: Combiner[Input, Input2],
   ): Endpoint[PathInput, combiner.Out, Err, Output, Auth] =
-    copy(input = input ++ (HttpCodec.content[Input2](name) ?? doc))
+    inCodec(HttpCodec.content[Input2](name) ?? doc)
 
-  def in[Input2: HttpContentCodec](mediaType: MediaType)(implicit
+  def in[Input2: EnvironmentTag: HttpContentCodec](mediaType: MediaType)(implicit
     combiner: Combiner[Input, Input2],
   ): Endpoint[PathInput, combiner.Out, Err, Output, Auth] =
-    copy(input = input ++ HttpCodec.content[Input2](mediaType))
+    inCodec(HttpCodec.content[Input2](mediaType))
 
-  def in[Input2: HttpContentCodec](mediaType: MediaType, doc: Doc)(implicit
+  def in[Input2: EnvironmentTag: HttpContentCodec](mediaType: MediaType, doc: Doc)(implicit
     combiner: Combiner[Input, Input2],
   ): Endpoint[PathInput, combiner.Out, Err, Output, Auth] =
-    copy(input = input ++ (HttpCodec.content(mediaType) ?? doc))
+    inCodec(HttpCodec.content(mediaType) ?? doc)
 
-  def in[Input2: HttpContentCodec](mediaType: MediaType, name: String)(implicit
+  def in[Input2: EnvironmentTag: HttpContentCodec](mediaType: MediaType, name: String)(implicit
     combiner: Combiner[Input, Input2],
   ): Endpoint[PathInput, combiner.Out, Err, Output, Auth] =
-    copy(input = input ++ HttpCodec.content(name, mediaType))
+    inCodec(HttpCodec.content(name, mediaType))
 
-  def in[Input2: HttpContentCodec](mediaType: MediaType, name: String, doc: Doc)(implicit
+  def in[Input2: EnvironmentTag: HttpContentCodec](mediaType: MediaType, name: String, doc: Doc)(implicit
     combiner: Combiner[Input, Input2],
   ): Endpoint[PathInput, combiner.Out, Err, Output, Auth] =
-    copy(input = input ++ (HttpCodec.content(name, mediaType) ?? doc))
+    inCodec(HttpCodec.content(name, mediaType) ?? doc)
 
   /**
    * Returns a new endpoint derived from this one, whose request must satisfy
    * the specified codec.
    */
-  def inCodec[Input2](codec: HttpCodec[HttpCodecType.RequestType, Input2])(implicit
+  def inCodec[Input2: EnvironmentTag](codec: HttpCodec[HttpCodecType.RequestType, Input2])(implicit
     combiner: Combiner[Input, Input2],
-  ): Endpoint[PathInput, combiner.Out, Err, Output, Auth] =
-    copy(input = input ++ codec)
+  ): Endpoint[PathInput, combiner.Out, Err, Output, Auth] = {
+    val isOptional = EnvironmentTag[Input2] <:< EnvironmentTag[Option[?]]
+    println(s"==================== Is optional: $isOptional ==================== ")
+    copy(input = input ++ (if (isOptional) codec.optional.asInstanceOf[zio.http.codec.HttpCodec[zio.http.codec.HttpCodecType.RequestType,Input2]] else codec))
+  }
```

Comment:
I attempted a fix here but the result is that it makes fail the only test which was passing and the test which was failing still fails